src/Telemetry/Microsoft.DotNet.AzureDevOpsTimelineTests/AzureDevOpsTimelineTests.cs

```diff
@@ -0,0 +1,308 @@
+using FluentAssertions;
+using Microsoft.DotNet.Internal.AzureDevOps;
+using Microsoft.DotNet.Internal.Testing.Utility;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Internal;
+using Microsoft.Extensions.Logging;
+using Moq;
+using NUnit.Framework;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.DotNet.AzureDevOpsTimeline.Tests
+{
+    public class AzureDevOpsTimelineTests
+    {
+        private class TestData : IDisposable
+        {
+            public readonly AzureDevOpsTimeline Controller;
+            private readonly ServiceProvider _services;
+
+            public TestData(AzureDevOpsTimeline controller, ServiceProvider services)
+            {
+                Controller = controller;
+                _services = services;
+            }
+
+            public void Dispose()
+            {
+                _services?.Dispose();
+            }
+        }
+
+        private static TestData SetUp(
+            ITimelineTelemetryRepository timelineTelemetryRepository, 
```

Comment:
It feels like this should probably go the other way... the "Setup" would create all the mocks and return them on the TestData class, rather than forcing every test to manage them itself.

Reply:
The way I've done this in other tests is by including an optional override parameter for mocks that you need specifically for your test case. Otherwise, the Setup method handles all the default mocks. 

example: https://dnceng.visualstudio.com/internal/_git/dotnet-helix-service?path=%2Fsrc%2FServiceFabric%2FHelix%2FBuildResultAnalysisProcessor.Tests%2FBuildAnalysisTests.cs&version=GBmain&line=51&lineEnd=52&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

Reply:
If the TestData had the MockWhatever as a property, you could just muck with it after the Setup call.

Reply:
I considered most of this. One issue is that the mock data is almost entirely different for each test, so I didn't see a benefit to having a builder.

Reply:
The builder stuff is _mostly_ if you need to change how a particular thing is created/injected.  In this case, if the builder took the list of builds, it could put them in the mock it injects, and you wouldn't need to expose that mock out to every test.

Reply:
I may have taken it a bit far, and I don't love everything about this implementation (a future attempt will be better), but I think I've landed in an okay place that reduces overall opportunity for typos and accidents and improves readability of the actual tests.

Reply:
This has also given me a lot to think about as a technique in managing an otherwise complicated test setup. I'm grateful for any feedback!